apps/opik-backend/src/main/java/com/comet/opik/domain/ProxyService.java

```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public ProviderApiKey get(UUID id) {
+        String workspaceId = requestContext.get().getWorkspaceId();
+
+        log.info("Getting provider api key with id '{}', workspaceId '{}'", id, workspaceId);
+
+        var providerApiKey = template.inTransaction(READ_ONLY, handle -> {
+
+            var repository = handle.attach(ProviderApiKeyDAO.class);
+
+            return repository.fetch(id, workspaceId).orElseThrow(this::createNotFoundError);
+        });
+        log.info("Got provider api key with id '{}', workspaceId '{}'", id, workspaceId);
+
+        return providerApiKey.toBuilder()
+                .apiKey(encryptionService.decrypt(providerApiKey.apiKey()))
+                .build();
+    }
+
+    @Override
+    public ProviderApiKey saveApiKey(@NonNull ProviderApiKey providerApiKey) {
+        UUID apiKeyId = idGenerator.generateId();
+        String userName = requestContext.get().getUserName();
+        String workspaceId = requestContext.get().getWorkspaceId();
+
+        var newProviderApiKey = providerApiKey.toBuilder()
+                .id(apiKeyId)
+                .apiKey(encryptionService.encrypt(providerApiKey.apiKey()))
+                .createdBy(userName)
+                .lastUpdatedBy(userName)
+                .build();
+
+        try {
+            template.inTransaction(WRITE, handle -> {
+
+                var repository = handle.attach(ProviderApiKeyDAO.class);
+                repository.save(workspaceId, newProviderApiKey);
+
+                return newProviderApiKey;
+            });
+
+            return get(apiKeyId);
+        } catch (UnableToExecuteStatementException e) {
+            if (e.getCause() instanceof SQLIntegrityConstraintViolationException) {
+                throw newConflict();
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public void updateApiKey(@NonNull UUID id, @NonNull ProviderApiKeyUpdate providerApiKeyUpdate) {
+        String userName = requestContext.get().getUserName();
+        String workspaceId = requestContext.get().getWorkspaceId();
+        String encryptedApiKey = encryptionService.encrypt(providerApiKeyUpdate.getApiKey());
+
+        template.inTransaction(WRITE, handle -> {
+
+            var repository = handle.attach(ProviderApiKeyDAO.class);
+
+            ProviderApiKey providerApiKey = repository.fetch(id, workspaceId)
+                    .orElseThrow(this::createNotFoundError);
+
+            repository.update(providerApiKey.id(),
+                    workspaceId,
+                    encryptedApiKey,
+                    userName);
+
+            return null;
+        });
+    }
+
+    private EntityAlreadyExistsException newConflict() {
+        log.info(PROVIDER_API_KEY_ALREADY_EXISTS);
+        return new EntityAlreadyExistsException(new ErrorMessage(List.of(PROVIDER_API_KEY_ALREADY_EXISTS)));
+    }
+
+    private NotFoundException createNotFoundError() {
```

Comment:
Minor: inherited from similar code, but this method name should be renamed in order to be reusable e.g: throwApiKeyNotFoundException or similar.

Reply:
I left as is, since this function is used as a `Supplier`, and it actually just creates an exception instance and doesn't throw anything.